packages/preferences-api/src/common/preference-service.ts

```diff
@@ -33,59 +39,64 @@ export class PreferenceService implements Disposable {
     }
 
     protected onDidChangePreference(event: PreferenceChangedEvent): void {
+        for (const prefChange of event.changes) {
+            if (prefChange.newValue === undefined || prefChange.newValue === null) {
+                delete this.prefCache[prefChange.preferenceName];
+            } else if (prefChange.newValue !== undefined) {
+                this.prefCache[prefChange.preferenceName] = prefChange.newValue;
+            }
+        }
+
+        this.resolveReady();
         this.onPreferenceChangedEmitter.fire(event);
     }
 
     get onPreferenceChanged(): Event<PreferenceChangedEvent> {
```

Comment:
it should be `Event<PreferenceChange>` here, fire all changes individually can be done once at line 51